### PR TITLE
4.14 backport: threads and fork

### LINKED ISSUE #12636: Possible corruption bug in `caml_master_lock` in 4.14.1 on macOS
A couple of days ago, I upgraded to macOS 14 Sonoma from 13 on my M1 Macbook Pro. After doing so, I noticed that my LSP server would seemingly randomly stop working and would spin, causing 100% CPU usage. 

I opened an issue (ocaml/ocaml-lsp#1194) thinking it was maybe a bug in the LSP server itself, but upon further investigation it appears that it might be a bug in the runtime.

Unfortunately, I haven't been able to find an easily reproducible example for this, but I'll continue to try to do so and outline my investigation here.

---

Once the LSP froze, I attached to it with lldb to see if I could glean some more info that way. Immediately after doing so, a breakpoint is hit (even though I hadn't set any):
```
* thread #1, stop reason = EXC_BREAKPOINT (code=1, subcode=0x188124220)
    frame #0: 0x0000000188124220 libsystem_platform.dylib`_os_unfair_lock_corruption_abort + 52
libsystem_platform.dylib`:
->  0x188124220 <+52>: brk    #0x1
    0x188124224 <+56>: stp    x20, x21, [sp, #-0x10]!
    0x188124228 <+60>: adrp   x20, 0
    0x18812422c <+64>: add    x20, x20, #0x95d          ; "BUG IN CLIENT OF LIBPLATFORM: os_unfair_lock is corrupt"
```

The backtrace at this point looks like the following:
```
* thread #1, stop reason = EXC_BREAKPOINT (code=1, subcode=0x188124220)
  * frame #0: 0x0000000188124220 libsystem_platform.dylib`_os_unfair_lock_corruption_abort + 52
    frame #1: 0x000000018811f788 libsystem_platform.dylib`_os_unfair_lock_lock_slow + 332
    frame #2: 0x00000001880ee3f0 libsystem_pthread.dylib`pthread_mutex_destroy + 64
    frame #3: 0x00000001008144fc ocamllsp`caml_thread_reinitialize [inlined] st_mutex_destroy(m=0x00006000016dc040) at st_posix.h:228:8 [opt]
    frame #4: 0x00000001008144f4 ocamllsp`caml_thread_reinitialize at st_stubs.c:425:7 [opt]
    frame #5: 0x00000001880f6f7c libsystem_pthread.dylib`_pthread_atfork_child_handlers + 76
    frame #6: 0x0000000187faeb90 libsystem_c.dylib`fork + 112
    frame #7: 0x000000010081323c ocamllsp`spawn_unix(v_env=<unavailable>, v_cwd=<unavailable>, v_prog=<unavailable>, v_argv=<unavailable>, v_stdin=<unavailable>, v_stdout=<unavailable>, v_stderr=<unavailable>, v_use_vfork=<unavailable>, v_setpgid=1) at spawn_stubs.c:439:43 [opt]
    frame #8: 0x0000000100849fc0 ocamllsp`caml_c_call + 28
    frame #9: 0x0000000100661e38 ocamllsp`camlSpawn__spawn_inner_997 + 152
    frame #10: 0x00000001003ba90c ocamllsp`camlOcaml_lsp_server__run_in_directory_5864 + 524
    frame #11: 0x000000010044e728 ocamllsp`camlMerlin_kernel__Pparse__apply_rewriter_786 + 296
    frame #12: 0x000000010044ec28 ocamllsp`camlMerlin_kernel__Pparse__rewrite_957 + 144
    frame #13: 0x000000010044eca4 ocamllsp`camlMerlin_kernel__Pparse__apply_rewriters_str_inner_1399 + 68
    frame #14: 0x000000010044ed74 ocamllsp`camlMerlin_kernel__Pparse__apply_rewriters_1089 + 52
    frame #15: 0x0000000100464564 ocamllsp`camlMerlin_kernel__Mppx__fun_1203 + 60
    frame #16: 0x0000000100464428 ocamllsp`camlMerlin_kernel__Mppx__code_begin + 136
    frame #17: 0x000000010044f81c ocamllsp`camlMerlin_kernel__Phase_cache__apply_inner_528 + 852
    frame #18: 0x0000000100465ca0 ocamllsp`camlMerlin_kernel__Mpipeline__fun_1778 + 152
    frame #19: 0x000000010064bdd0 ocamllsp`camlMerlin_utils__Misc__try_finally_inner_3715 + 48
    frame #20: 0x00000001007b993c ocamllsp`camlCamlinternalLazy__force_lazy_block_362 + 140
    frame #21: 0x000000010046482c ocamllsp`camlMerlin_kernel__Mpipeline__fun_1613 + 172
    frame #22: 0x00000001007b993c ocamllsp`camlCamlinternalLazy__force_lazy_block_362 + 140
    frame #23: 0x0000000100465d44 ocamllsp`camlMerlin_kernel__Mpipeline__fun_1791 + 60
    frame #24: 0x00000001007b993c ocamllsp`camlCamlinternalLazy__force_lazy_block_362 + 140
    frame #25: 0x000000010046482c ocamllsp`camlMerlin_kernel__Mpipeline__fun_1613 + 172
    frame #26: 0x00000001007b993c ocamllsp`camlCamlinternalLazy__force_lazy_block_362 + 140
    frame #27: 0x000000010042acb0 ocamllsp`camlQuery_commands__dispatch_2992 + 7320
    frame #28: 0x0000000100646d24 ocamllsp`camlMerlin_utils__Std__let_ref_3289 + 60
    frame #29: 0x000000010064bdd0 ocamllsp`camlMerlin_utils__Misc__try_finally_inner_3715 + 48
    frame #30: 0x00000001007f8048 ocamllsp`camlStdlib__Fun__protect_317 + 96
    frame #31: 0x000000010045acd8 ocamllsp`camlMerlin_kernel__Mocaml__with_state_438 + 80
    frame #32: 0x000000010037dd44 ocamllsp`camlOcaml_lsp_server__Document__fun_4117 + 116
    frame #33: 0x00000001007a1d18 ocamllsp`camlStdune__Exn_with_backtrace__try_with_422 + 40
    frame #34: 0x00000001006645c8 ocamllsp`camlLev_fiber__do_no_raise_1549 + 24
    frame #35: 0x000000010066fd38 ocamllsp`camlLev_fiber_util__Worker__run_322 + 88
    frame #36: 0x0000000100671354 ocamllsp`camlThread__fun_850 + 44
    frame #37: 0x000000010084a030 ocamllsp`caml_start_program + 104
    frame #38: 0x000000010083ffa0 ocamllsp`caml_callback_exn(closure=<unavailable>, arg=1) at callback.c:111:10 [opt]
    frame #39: 0x0000000100814768 ocamllsp`caml_thread_start(arg=<unavailable>) at st_stubs.c:549:5 [opt]
    frame #40: 0x00000001880f3034 libsystem_pthread.dylib`_pthread_start + 136
```

This seems to refer to code similar to this found in `libplatform` in the version of Darwin used by macOS 14: https://github.com/apple/darwin-libplatform/blob/215b09856ab5765b7462a91be7076183076600df/src/os/lock.c#L136

Dumping the register state here is also a bit informative:
```
General Purpose Registers:
        x0 = 0x00000000000007fa
        x1 = 0x0000000000000000
        x2 = 0x00000000000007fa
        x3 = 0x0000000000000000
        x4 = 0x00000d5200000000
        x5 = 0x0000040300000001
        x6 = 0x0000000000002000
        x7 = 0x0000000000000000
        x8 = 0x00000000000007fa
        x9 = 0x00000000000005fa
       x10 = 0x000000000003fe00
       x11 = 0x0000000100e33a38  ocamllsp`caml_master_lock + 24
       x12 = 0x0000000100e33a38  ocamllsp`caml_master_lock + 24
       x13 = 0x00000000ffff8128
       x14 = 0x0000000000000010
       x15 = 0x00000000ffff7dff
       x16 = 0x0000000000000203
       x17 = 0x00000001e85e8e58
       x18 = 0x0000000000000000
       x19 = 0x0000000000000303
       x20 = 0x0000000000050000
       x21 = 0x00006000016dc048
       x22 = 0x0000000001050002
       x23 = 0x00000000000007fa
       x24 = 0x0000000000000000
       x25 = 0x00000000ffffffff
       x26 = 0x0000000000000303
       x27 = 0x0000000000000000
       x28 = 0x000000016fc367b0
        fp = 0x000000016fc36610
        lr = 0x000000018811f788  libsystem_platform.dylib`_os_unfair_lock_lock_slow + 332
        sp = 0x000000016fc365d0
        pc = 0x0000000188124220  libsystem_platform.dylib`_os_unfair_lock_corruption_abort + 52
      cpsr = 0x80001000
```

It's unclear to me what specifically `x11` and `x12` are being used for here, but they both contain pointers inside `caml_master_lock`, which libplatform seems to believe the lock inside is corrupt. 

The backtrace also makes mention of `caml_thread_reinitialize`, which appears to be calling `st_mutex_destroy` over all open channels:
https://github.com/ocaml/ocaml/blob/49bff4cb1597b095ba5b455b989f456ff6939e76/otherlibs/systhreads/st_stubs.c#L421-L428

Unfortunately that's sort of where my investigation has ended -- I can't seem to figure out _if_ the lock is actually even corrupt or not, let alone where that's happening if so. 

If anyone has any pointers for where I should look next, please let me know. I'll work on making a reproducible example.

### PR DESCRIPTION
**Pull request by xavierleroy**
This is a backport to 4.14 of some of the changes in system threads related to the `fork` system call:
- Use a `caml_atfork_hook` instead of `pthread_atfork` to reinitialize the thread machinery after `Unix.fork` (backport of c6d00ee0d44555adb2ec1ce86e8519beff1182c2)
- Remove remaining code related to `pthread_atfork` (backport of 650acfccfc2f09ef88f5acded0e04715e69d6d73)
- Don't destroy locks on I/O channels, this may be dangerous (backport of a change in 5.0 that traces back to Multicore OCaml and that I failed to identify).

Could possibly fix #12636 (to be tested if at all possible).

### CODE REVIEW

**Review by abbysmal — APPROVED** [review]
Changes are pretty straightforward and just a back port of 5.0 changes: looks good to me.

Considering 5.0's mileage of running problematic programs (like `dune` or `ocaml-lsp`, both exhibiting the same after fork behaviour), I don't think we will see a massive regression here.

We can merge this as is, or maybe wait for a feedback from the reporter at #12636 (though this fix is correct, as destroying potentially locked mutexes is wrong.)

**Comment by gadmm** [comment]
The solution for 5.x is therefore to remove the `FIXME` comment about this, or is there another comment/course of action?

**Comment by xavierleroy** [comment]
> this fix is correct, as destroying potentially locked mutexes is wrong

But leaving them locked forever is wrong as well :-(   The only standard-conforming implementation that I know of is to lock all the I/O mutexes before forking, then unlock them in both the parent and child processes.  But that can block `Unix.fork` for a long time, waiting for I/O to complete.

A middle ground would be to reinitialize the I/O mutexes in the child process.  This is still nonstandard behavior but perhaps less dangerous than destroying them.  Amusingly, the code started with reinitialization, and switched to destruction in commit https://github.com/ocaml/ocaml/commit/09ba69134f63132a1516dc57c7d17a9c8b1475d9 , and I have no idea why.

**Comment by abbysmal** [comment]
Sorry for the late reply.

> A middle ground would be to reinitialize the I/O mutexes in the child process.

This sounds reasonable (glossing over glibc implementation at least, which is not the source of the reported problem anyway.).

Now the question is how consistent are implementation of `pthreads_mutex_init` over cleaning up, and could some trigger assertions if attempting to initialise an already initialised mutex?

TBQH, I think the "not caring about cleaning IO mutexes" behaviour is fine. The runtime already offers support for a highly discouraged practice, in a best effort fashion.

There are other ways to go about it: we could document that people should be careful when `forking` with threads to make sure no thread is holding a lock over a channel, this would ensure the child process can _probably_ destroy them properly.

Or, the right answer, would be to just not tempt the devil by using `fork` in a multithreaded context and write correct code. But this may be naive and I may be missing legitimate use cases for this.

(Sorry for the grumpy tone, I did spend a lot of time working on `dune` running on the early Multicore builds, this brought back not so happy memories :-))

**Comment by gasche** [comment]
> We could document that people should be careful when forking with threads to make sure no thread is holding a lock over a channel, this would ensure the child process can probably destroy them properly.

I don't see how users would be able to respect this recommendation if they do use `fork` in a multi-threaded programs. Most I/O operations can schedule a thread out with the mutex locked, the user cannot detect this and I don't see what programming style they could use to reason about whether channels are locked by other threads when they want to `fork()`.

> Or, the right answer, would be to just not tempt the devil by using `fork` in a multithreaded context and write correct code. But this may be naive and I may be missing legitimate use cases for this.

I agree that avoiding a mix of `fork()` and threads is probably the wisest course of action. In older versions of OCaml we would use `fork()` to implement `exec*`-family functions, and this usage has persisted in the ecosystem. (I guess sometimes it still has uses, if you want some cleanup logic to run on the child process before executing.)

**Comment by abbysmal** [comment]
> I don't see how users would be able to respect this recommendation if they do use fork in a multi-threaded programs.

Sorry, my statement was a bad attempt at being tongue-in-cheek, it is indeed rather hard (if not impossible) in the most common threading scenarios in OCaml programs. :-)

> (I guess sometimes it still has uses, if you want some cleanup logic to run on the child process before executing.)

Do we have any report of such use cases?

**Comment by xavierleroy** [comment]
> >    (I guess sometimes it still has uses, if you want some cleanup logic to run on the child process before executing.)

> Do we have any report of such use cases?

If you look at Jane Street's "spawn" library (https://github.com/janestreet/spawn), you'll see that it does many things in the child process before calling exec.  A good example is to change the working directory for the program that will be exec'd.  It happens that they do all of this (fork, chdir, exec) in C, so once this PR is merged, OCaml's support for fork and threads will no longer interfere with the "spawn" library.  But if someone wanted to write the same kind of custom launcher in OCaml, they would still have to use Unix.fork.

**Comment by xavierleroy** [comment]
More generally: in discussing this PR and fork/thread related issues, please distinguish between calling fork() from C and calling Unix.fork () from OCaml.

**Comment by gasche** [comment]
Yes, I had this same `chdir` example in mind.

This is a minor point, but I am not fully convinced by the distinction between forking from OCaml and forking from C. I think of `Unix.fork()` as something that will fork *and* ensure that the OCaml runtime system is in a consistent state after forking. If I am calling `fork()` from C, from an OCaml program, and I really want a fork (not exec after a short time), I want to run the same logic to be able to continue executing OCaml code on the child. (It is fine if I need to call a runtime-provided function on the child to do this, instead of having pthread do it magically for me.)

Here it is not just "a `fork()` from C", but more importantly "a `fork()` that will not rely on the OCaml runtime anymore". So I think that the real distinction would be whether we want to "restore" the OCaml runtime on the child, or we are giving up on accessing the runtime. In fact, it would make sense to have an explicit function to call on the child to "give up", instead of hoping that "doing nothing" is always a correct way to do it.

My understanding of the issue at hand is that the 4.x runtime assumes that all `fork()` calls (from C) in an OCaml program want to "restore" the runtime state on the child, and makes implementation choices that are wrong if the intent was to "give up" -- because they error out in cases where restoring is basically impossible but giving up is safe.

**Comment by xavierleroy** [comment]
I got cold feet about leaving I/O mutexes possibly locked in the child process, as this is a regression w.r.t. 4.14, so I pushed a new version of this PR that reinitializes the mutexes instead of destroying them.

Looking at the sources for Apple's pthread library (https://github.com/apple-oss-distributions/libpthread/tree/libpthread-519), it is clear that pthread_mutex_init doesn't try to check for an already-initialized, already-locked mutex, while pthread_mutex_destroy starts by acquiring the underlying OS lock, and this seems to be causing the fatal error reported in #12636.

> Amusingly, the code started with reinitialization, and switched to destruction in commit https://github.com/ocaml/ocaml/commit/09ba69134f63132a1516dc57c7d17a9c8b1475d9 , and I have no idea why.

I figured it out!  This commit is the one that introduced the st_posix.h and st_win32.h interfaces, instead of calling directly into POSIX or Win32 functions.  These interfaces provide an "allocate and initialize mutex" operation, and a "destroy and free mutex" operation, but no "reinitialize already-allocated mutex" operation.  This PR adds the latter operation; it is a no-op for Win32 since there is no fork() and therefore no use case.

**Comment by xavierleroy** [comment]
> My understanding of the issue at hand is that the 4.x runtime assumes that all fork() calls (from C) in an OCaml program want to "restore" the runtime state on the child, and makes implementation choices that are wrong if the intent was to "give up" -- because they error out in cases where restoring is basically impossible but giving up is safe.

That's a good way of looking at it.

**Comment by xavierleroy** [comment]
@Engil are you still happy with this PR, now that I/O mutexes are reinitialized after fork?

**Comment by abbysmal** [comment]
@xavierleroy Looks good to me, as a bonus it seems to be fixing [the problem witnessed on recent MacOS versions + LSP](https://github.com/ocaml/ocaml/issues/12636#issuecomment-1769070360).

Let's merge it. (need Changes entry fixup)

> These interfaces provide an "allocate and initialize mutex" operation, and a "destroy and free mutex" operation, but no "reinitialize already-allocated mutex" operation. This PR adds the latter operation; it is a no-op for Win32 since there is no fork() and therefore no use case.

It make sense to me, thank you for the digging through the past. 🙂 

> The solution for 5.x is therefore to remove the FIXME comment about this, or is there another comment/course of action?

I think as such the 5.X comment is still pretty much valid and doesn't need to be touched.

**Comment by xavierleroy** [comment]
Thanks for the feedback.  Now merging!

**Pull request merged**
